DigitalLearningSolutions.Data.Tests/Services/SessionServiceTests.cs

```diff
@@ -1,152 +1,99 @@
-                );
-            }
-        }
+            httpContextSession = new MockHttpContextSession();
 
-        [Test]
-        public void StartOrRestartSession_Should_Start_Subsequent_Session()
-        {
-            using (new TransactionScope())
-            {
-                // When
-                const int candidateId = 101;
-                const int customisationId = 1240;
-                var sessionId = sessionService.StartOrRestartSession(candidateId, customisationId);
-
-                // Then
-                var newSession = sessionTestHelper.GetSession(sessionId);
-
-                newSession.Should().NotBeNull();
-                SessionTestHelper.SessionsShouldBeApproximatelyEquivalent(
-                    newSession!,
-                    SessionTestHelper.CreateDefaultSession(sessionId)
-                );
-            }
+            sessionService = new SessionService(sessionDataService);
         }
 
         [Test]
-        public void StartOrRestartSession_Should_Stop_Other_Sessions()
+        public void Index_should_StartOrRestartSession_for_course_not_in_session()
         {
-            using (new TransactionScope())
-            {
-                // When
-                const int candidateId = 101;
-                const int customisationId = 1240;
-                var sessionId = sessionService.StartOrRestartSession(candidateId, customisationId);
-
-                // Then
-                var sessions = sessionTestHelper.GetCandidateSessions(candidateId);
-                var activeSessions = sessions.Where(session => session.Active);
-
-                activeSessions.Should().HaveCount(1);
-                activeSessions.First().SessionId.Should().Be(sessionId);
-            }
+            // Given
+            httpContextSession.Clear();
+            const int newCourseId = CustomisationId;  // Not in session
+            const int oldCourseInSession = CustomisationId + 1;
+            httpContextSession.SetInt32($"SessionID-{oldCourseInSession}", DefaultSessionId + 1);
+
+            // When
+            sessionService.StartOrUpdateSession(CandidateId, newCourseId, httpContextSession);
+
+            // Then
+            A.CallTo(() => sessionDataService.StartOrRestartSession(CandidateId, newCourseId)).MustHaveHappenedOnceExactly();
+            A.CallTo(() => sessionDataService.StartOrRestartSession(A<int>._, A<int>._))
+                .WhenArgumentsMatch((int candidateId, int customisationId) =>
+                    candidateId != CandidateId || customisationId != newCourseId)
+                .MustNotHaveHappened();
+
+            A.CallTo(() => sessionDataService.UpdateSessionDuration(A<int>._)).MustNotHaveHappened();
+
+            httpContextSession.Keys.Should().BeEquivalentTo($"SessionID-{newCourseId}");
+            httpContextSession.GetInt32($"SessionID-{newCourseId}").Should().Be(DefaultSessionId);

```

Comment:
I'd still split this into two tests; one for checking the correct methods are being called on the session data service, one for checking the http context session has been updated 🙂

Reply:
Yeah that looks like a better approach. Refactored in 90a4212 🙂